Wordlist View

shakil-fg opened the pull request.

chuanhtuan commented on app/assets/javascripts/lessons.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
remove unused file

chuanhtuan commented on app/assets/stylesheets/lessons.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Lessons controller here.
```
remove unused file

chuanhtuan commented on config/routes.rb:
```diff
@@ -1,4 +1,14 @@
 Rails.application.routes.draw do
+  get 'wordlist'  => 'lessons#wordlist'
```
please use word controller

chuanhtuan commented on config/routes.rb:
```diff
@@ -1,4 +1,14 @@
 Rails.application.routes.draw do
+  get 'wordlist'  => 'lessons#wordlist'
+
+  get 'lessons/category'
```
define a resources lesson

chuanhtuan commented on app/controllers/lessons_controller.rb:
```diff
@@ -0,0 +1,10 @@
+  def show
+  end
+
+  def result
```
remove this method
RESTful

shakil-fg replied:
I need "show" action for showing lesson questions. I also need "result" action to show the result. What action should if I remove "result" action?

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,66 @@
+
+  def load_dependencies
+    @user = current_user
+    @categories = category_list
```
@categories = Category.all

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,66 @@
+  end
+
+  private
+  def category_list
```
remove this method

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,66 @@
+    wlst={}
+    if @word_filter["word_stat"] == "not_learned"
+      learned_words =[]
+      lessons = @user.lessons.where(category_id: Category.find_by(name: @word_filter["ctgs"]))
```
any "where clause" have to write in a scope

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,37 @@
+    <p><%= t :ppuyi %></p>
+<% end %>
+
+<%= form_for(:lesson, url: wordlist_path) do |f| %>
```
should use form_tag
url: words_path

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,62 @@
+    @category_name = Category.find(params[:ctgs].to_i).name
+  end
+
+  def generate_wordlist
```
this method is too long
you only need 3 scope in model
1. load all
2. learned
3. not_learned

shakil-fg replied:
I didn't understand the point. Could you please explain?

shakil-fg replied:
I apologize for asking again.

I don't understand how to implement scope here. Let me explain where I faced difficulties.

You mentioned “scope load all word”. I need to know the category to load all words in that category. So, my query is
Category.find(params[:ctgs].to_i).words.each do |wd|
How can I short it down?
I need to load word answer along with it. So, I used the query
wlst[wd.content] = wd.word_answers.where(correct: true).first.content
	How can I short it down?
You mentioned, “scope load learned word”. I need to load learned words for a user from the selected category. I used
		@user.lessons & Category.find(params[:ctgs].to_i).lessons
What scope can be implemented here?

I am using scope for the first time. Please give me some reference for it for gaining a better understanding and knowing about the usage scenario.

shakil-fg commented:
I apologize for asking again.

I don't understand how to implement scope here. Let me explain where I faced difficulties.

You mentioned “scope load all word”. I need to know the category to load all words in that category. So, my query is
Category.find(params[:ctgs].to_i).words.each do |wd|
How can I short it down?
I need to load word answer along with it. So, I used the query
wlst[wd.content] = wd.word_answers.where(correct: true).first.content
	How can I short it down?
You mentioned, “scope load learned word”. I need to load learned words for a user from the selected category. I used
		@user.lessons & Category.find(params[:ctgs].to_i).lessons
What scope can be implemented here?

I am using scope for the first time. Please give me some reference for it for gaining a better understanding and knowing about the usage scenario.

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  def index
+  end
+
+  def create
```
remove

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+<%= form_tag({controller: :word, action: :create}, method: :post) do %>
+    <div>
+      <%= label_tag(:ctgs, t(:lsncat)) %>
+      <%= select_tag(:ctgs, options_for_select(@categories.map { |category| [ category.name, category.id ] })) %>
```
<%= select_tag :category_id, options_from_collection_for_select(@categories, :id, :name)

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  def load_dependencies
+    @user = current_user
+    @categories = Category.all
+    @category_name = Category.find(params[:ctgs].to_i).name unless params[:ctgs].nil?
```
@category = Category.find params[:category_id] unless params[:category_id].nil?

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def generate_wordlist
+    wlst={}
```
Write scope in Word model
learned_word_ids = "SELECT word_id FROM results JOIN lessons ON lessons.id = results.lesson_id  WHERE lessons.user_id = :user_id"
scope :learned_words, ->user {where "id IN (#{learned_word_ids})", user_id: user.id}
scope :not_learned_words, ->user {where "id NOT IN (#{learned_word_ids})",
    user_id: user.id}
scope :filter_category, ->category_id {where category_id: category_id if category_id.present?}

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+    <p><%= t :ppuyi %></p>
+<% end %>
+
+<%= form_tag({controller: :word, action: :create}, method: :post) do %>
```
<%= form_tag words_path, method: :get do %>

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+      <%= label_tag(:ctgs, t(:lsncat)) %>
+      <%= select_tag(:ctgs, options_for_select(@categories.map { |category| [ category.name, category.id ] })) %>
+
+      <%= label_tag(:word_stat, t(:wdstat)) %>
```
()

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+      <%= select_tag(:ctgs, options_for_select(@categories.map { |category| [ category.name, category.id ] })) %>
+
+      <%= label_tag(:word_stat, t(:wdstat)) %>
+      <%= radio_button_tag(:word_stat, :learned) %>
```
()

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+    </div>
+<% end %>
+
+<% if !(@wordlist.nil?) && @wordlist.count > 0 %>
```
<% unless @wordlist.blank? %>

chuanhtuan commented on config/routes.rb:
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  get 'word'  => 'word#index'
```
resources :words, only: [:index]

chuanhtuan commented on app/controllers/word_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def generate_wordlist
+    wlst={}
```
if params[:word_stat] == "not_learned"
  @words = Word.not_learned_words(@user).filter_category(params[:category_id])
elsif params[:word_stat] == "learned"
  @words = Word.learned_words(@user).filter_category(params[:category_id])
else
  @words = Word.filter_category(params[:category_id]).words
end

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+    <p><%= t :ppuyi %></p>
+<% end %>
+
+<%= form_tag word_index_path, method: :get do %>
```
<%= form_tag words_path, method: :get do %>

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+<% provide(:title, t(:wordlist)) %>
+
+<% if !(params[:ctgs].nil?) %>
```
params[:category_id]

chuanhtuan commented on app/models/word.rb:
```diff
@@ -5,4 +5,9 @@ class Word < ActiveRecord::Base
 
   validates :content,  presence: true
+
+  learned_word_ids = "SELECT words.id FROM lessons_words, words, word_answers, lessons WHERE lessons.id = lessons_words.lesson_id AND words.id = lessons_words.word_id AND words.id=word_answers.word_id AND word_answers.correct=\"t\""
```
no need to check correct
because this query will run very slow
you full join form 4 tables

shakil-fg replied:
I tried the following test Query
"SELECT * FROM words FULL JOIN word_answers ON words.id = word_answers.word_id" 
But I got following error
SQLException: RIGHT and FULL OUTER JOINs are not currently supported

chuanhtuan commented on app/views/word/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+        <td><%= t :word %></td>
+        <td><%= t :equivalent %></td>
+      </tr>
+      <% @wordlist.each do |x,y| %>
```
<% @word.each do |word| %>
<tr>
  <td><%= word.content %></td>
  <td><%= word.word_answers.correct_content %></td>
</tr>

dieunb commented on app/controllers/words_controller.rb:
```diff
@@ -0,0 +1,34 @@
+
+  private
+  def load_dependencies
+    @user = current_user
```
You can use `current_user` on controller and view, so no need set it to new variable.

shakil-fg replied:
If I do not set
@user = current_user
I'll have to write statement
Word.not_learned_words(@user).filter_category(@category.id)
as
Word.not_learned_words(current_user).filter_category(@category.id)

That will execute the entire current_user method. I have two usage of @user in words_controller. If I do not set @user, this will execute current_user twice. I think this is inefficient.

dieunb replied:
yep, why not. you can use it every case.

dieunb commented on app/controllers/words_controller.rb:
```diff
@@ -0,0 +1,34 @@
+class WordsController < ApplicationController
+  before_action :load_dependencies, :logged_in_user
```
in the order function, `load_dependencies` will call first and the next `logged_in_user`. For the case, user don't login,
actually it can have no meaning.

dieunb commented on app/models/category.rb:
```diff
@@ -3,4 +3,7 @@ class Category < ActiveRecord::Base
 
   validates :name,  presence: true
+
+  scope :lessons_in, -> (category) { find(category).lessons }
```
Not necessary, you can get lessons in category by `category.lessons`. I do as it because there is relation `1-n` between category and lesson

dieunb commented on app/models/category.rb:
```diff
@@ -3,4 +3,7 @@ class Category < ActiveRecord::Base
   validates :name,  presence: true
+
+  scope :lessons_in, -> (category) { find(category).lessons }
+  scope :words_in, -> (category) { find(category).words }
```
the same above

dieunb commented on app/models/word_answer.rb:
```diff
@@ -3,4 +3,7 @@ class WordAnswer < ActiveRecord::Base
 
   validates :content,  presence: true
+
+  scope :correct_one, -> { where(correct: true) }
```
the simple `scope :correct -> {where correct: true}`

dieunb commented on app/models/word_answer.rb:
```diff
@@ -3,4 +3,7 @@ class WordAnswer < ActiveRecord::Base
   validates :content,  presence: true
+
+  scope :correct_one, -> { where(correct: true) }
+  scope :correct_content, -> { correct_one.first.content }
```
no need.

shakil-fg replied:
It's used in "words/index.html.erb", on line 37

<td><%= word.word_answers.correct_content %></td>

dieunb commented on app/views/words/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+<% provide(:title, t(:wordlist)) %>
+
+<% if !(params[:category_id].nil?) %>
+    <h1><%= @category.name %> - <%= t :wordlist %></h1>
```
indent

dieunb commented on app/views/words/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+<% provide(:title, t(:wordlist)) %>
```
remove `(0`

dieunb commented on app/views/words/index.html.erb:
```diff
@@ -0,0 +1,41 @@
+<% provide(:title, t(:wordlist)) %>
+
+<% if !(params[:category_id].nil?) %>
```
why not `if params[:category_id].present?`

shakil-fg commented:
Sir, I don't have user_id field in lessons_words table (your query returned that user_id don't exist). So, I modified the query into

"SELECT words.id FROM lessons_words, lessons WHERE lessons.user_id = :user_id AND lessons.id = lessons_words.lesson_id"

Sir, one thing I am not clear about, 
Which words should fall in learned words category?
	Option 1: Words that user encountered during any lesson, regardless of whether s/he chosen the right answer or not
	Option 2: Words that  user encountered during any lesson and answered correctly (correct: true)

I followed option 2. But according to your query, I cannot use the correctness checking.

NOTE: I implemented your mentioned query (Option 1) in  this pull request
